handlers/whatsapp/whatsapp.go

```diff
@@ -593,25 +658,106 @@ func (h *handler) SendMsg(ctx context.Context, msg courier.Msg) (courier.MsgStat
 			externalID := ""
-			for i, part := range parts {
-				payload := mtTextPayload{
-					To:   msg.URN().Path(),
-					Type: "text",
-				}
-				payload.Text.Body = part
-				wppID, externalID, logs, err = sendWhatsAppMsg(msg, sendPath, payload)
 
-				// add logs to our status
-				for _, log := range logs {
-					status.AddLog(log)
+			qrs := msg.QuickReplies()
+
+			channel := msg.Channel()
+
+			wppVersion := channel.ConfigForKey("version", "0").(string)
+			hasCompatibleVersion := wppVersion >= interactiveMsgMinSupVersion
+
+			if len(qrs) > 0 && hasCompatibleVersion {
+				if len(qrs) > 3 {
+					section := mtSection{
+						Rows: make([]mtSectionRow, len(qrs)),
+					}
+
+					for i, qr := range qrs {
+						section.Rows[i] = mtSectionRow{
+							ID:    fmt.Sprint(i),
+							Title: qr,
+						}
+					}
+
+					for i, part := range parts {
+						payload := mtInteractivePayload{
+							To:   msg.URN().Path(),
+							Type: "interactive",
+						}
+						payload.Interactive.Type = "list"
+						payload.Interactive.Body.Text = part
+						payload.Interactive.Action.Button = "Menu"
+						payload.Interactive.Action.Sections = []mtSection{
+							section,
+						}
+
+						wppID, externalID, logs, err = sendWhatsAppMsg(msg, sendPath, payload)
+
+						for _, log := range logs {
+							status.AddLog(log)
+						}
+						if err != nil {
+							break
+						}
+
+						if i == 0 {
+							status.SetExternalID((externalID))
+						}
+					}
+
+				} else {
+					btns := make([]mtButton, len(qrs))
+					for i, qr := range qrs {
+						btns[i] = mtButton{
+							Type: "reply",
+						}
+						btns[i].Reply.ID = fmt.Sprint(i)
+						btns[i].Reply.Title = qr
+					}
+
+					for i, part := range parts {
+						payload := mtInteractivePayload{
+							To:   msg.URN().Path(),
+							Type: "interactive",
+						}
+						payload.Interactive.Type = "button"
+						payload.Interactive.Body.Text = part
+						payload.Interactive.Action.Buttons = btns
+						wppID, externalID, logs, err = sendWhatsAppMsg(msg, sendPath, payload)
+
+						for _, log := range logs {
+							status.AddLog(log)
+						}
+						if err != nil {
+							break
+						}
+
+						if i == 0 {
+							status.SetExternalID((externalID))
+						}
+					}
 				}
-				if err != nil {
-					break
-				}
-
-				// if this is our first message, record the external id
-				if i == 0 {
-					status.SetExternalID(externalID)
+			} else {
+				for i, part := range parts {
+					payload := mtTextPayload{
+						To:   msg.URN().Path(),
+						Type: "text",
+					}
+					payload.Text.Body = part
+					wppID, externalID, logs, err = sendWhatsAppMsg(msg, sendPath, payload)
+
+					// add logs to our status
+					for _, log := range logs {
+						status.AddLog(log)
+					}
+					if err != nil {
+						break
+					}
+
+					// if this is our first message, record the external id
+					if i == 0 {
+						status.SetExternalID(externalID)
+					}
```

Comment:
A lot of repetition above, can you refactor this so we aren't repeating `sendWhatsAppMsg`, logging and external id over and over? Maybe it's a callback function that's passed in to build the payload?

Reply:
I think your suggestion makes a lot of sense. So I tried to improve it that way, what did you think of how it is now? any suggestions are welcome and i will always be willing to try to improve.